osutil/group.go

```diff
@@ -20,35 +20,112 @@
 package osutil
 
 import (
+	"bytes"
+	"fmt"
+	"os/exec"
 	"os/user"
 	"strconv"
 )
 
-// TODO: the builtin os/user functions only look at /etc/passwd and /etc/group
-// which is fine for our purposes today. In the future we may want to support
-// lookups in extrausers, which is configured via nsswitch.conf. Since snapd
-// does not support being built with cgo itself, when we want to support
-// extrausers here, we can convert these to do the equivalent of:
-//
-//   getent passwd <user> | cut -d : -f 3
-//   getent group <group> | cut -d : -f 3
-
-// FindUid returns the identifier of the given UNIX user name.
-func FindUid(username string) (uint64, error) {
-	user, err := user.Lookup(username)
+var (
+	FindUid       = findUid
+	FindGid       = findGid
+	FindUidGetent = findUidGetent
+	FindGidGetent = findGidGetent
+)
+
+// The builtin os/user functions only look at /etc/passwd and /etc/group and
+// nothing configured via nsswitch.conf, like extrausers. findUid() and
+// findGid() continue this behavior where findUidGetent() and findGidGetent()
+// will perform a 'getent <database> <name>'
+
+// findUid returns the identifier of the given UNIX user name with no getent
+// fallback
+func findUid(username string) (uint64, error) {
+	myuser, err := user.Lookup(username)
 	if err != nil {
 		return 0, err
 	}
 
-	return strconv.ParseUint(user.Uid, 10, 64)
+	return strconv.ParseUint(myuser.Uid, 10, 64)
 }
 
-// FindGid returns the identifier of the given UNIX group name.
-func FindGid(groupname string) (uint64, error) {
+// findGid returns the identifier of the given UNIX group name with no getent
+// fallback
+func findGid(groupname string) (uint64, error) {
 	group, err := user.LookupGroup(groupname)
 	if err != nil {
 		return 0, err
 	}
 
 	return strconv.ParseUint(group.Gid, 10, 64)
 }
+
+// getent returns the identifier of the given UNIX user or group name as
+// determined by the specified database
+func getent(name string, database string) (uint64, error) {
+	if database != "passwd" && database != "group" {
+		return 0, fmt.Errorf(`unsupported getent database "%q"`, database)
+	}
+
+	cmdStr := []string{
+		"getent",
+		database,
+		name,
+	}
+	cmd := exec.Command(cmdStr[0], cmdStr[1:]...)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		// The API doesn't support return codes for exec.Command,
+		// so assume not found (which is reasonable since 'man getent'
+		// says that the other return codes are for unknown databases
+		// and enumeration not supported. 'passwd' will always exist
+		// and we aren't enumerating).
```

Comment:
I agree with the analysis. But one can get the error codes, by checking if err is a exec.ExitError I think and going from there.

Reply:
Thanks @mvo5 for fixing this!